torchdata/datapipes/iter/transform/callable.py

```diff
@@ -205,3 +206,162 @@ def __len__(self) -> int:
         if isinstance(self.datapipe, Sized):
             return len(self.datapipe)
         raise TypeError(f"{type(self).__name__} instance doesn't have valid length")
+
+
+@functional_datapipe("islice")
+class ISliceIterDataPipe(IterDataPipe[T_co]):
+    r"""
+    returns a slice of elements in input DataPipe via start/stop/step or indices (functional name: ``islice``).
+
+    Args:
+        datapipe: IterDataPipe with iterable elements
+        index: a single start index for the slice or a list of indices to be returned instead of a start/stop slice
+        stop: the slice stop. ignored if index is a list
+        step: step to be taken from start to stop. ignored if index is a list
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper([(0, 10, 100), (1, 11, 111), (2, 12, 122), (3, 13, 133), (4, 14, 144)])
+        >>> islice_dp = dp.islice(0, 2)
+        >>> list(islice_dp)
+        [(0, 10), (1, 11), (2, 12), (3, 13), (4, 14)]
+    """
+    datapipe: IterDataPipe
+
+    def __init__(
+        self,
+        datapipe: IterDataPipe,
+        index: Union[int, List[Hashable]],
+        stop: Optional[int] = None,
+        step: Optional[int] = None,
+    ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+
+        self.index = index
+        self.stop = stop
+        self.step = step
+
+    def __iter__(self) -> Iterator[T_co]:
+        for old_item in self.datapipe:
+            if isinstance(old_item, tuple):
+                if isinstance(self.index, list):
+                    new_item = tuple(x for i, x in enumerate(old_item) if i in self.index)  # type: ignore[assignment]
+                else:
+                    new_item = old_item[self.index : self.stop : self.step]  # type: ignore[assignment]
+            elif isinstance(old_item, list):
+                if isinstance(self.index, list):
+                    new_item = [x for i, x in enumerate(old_item) if i in self.index]  # type: ignore[assignment]
+                else:
+                    new_item = old_item[self.index : self.stop : self.step]  # type: ignore[assignment]
+            elif isinstance(old_item, dict):
+                if isinstance(self.index, list):
+                    new_item = {k: v for (k, v) in old_item.items() if k in self.index}  # type: ignore[assignment]
+                else:
+                    new_keys = list(old_item.keys())[self.index : self.stop : self.step]
+                    new_item = {k: v for (k, v) in old_item.items() if k in new_keys}  # type: ignore[assignment]
```

Comment:
For the case of `dict`, could we align the behavior no matter `self.index` is list or slice? TBH, supporting slice for `dict` is a little bit weird to me.

Reply:
that's fair. dict is ordered nowadays by default, so it seemed ok to have a proper slice out of it with that expectation. I'm also fine with just having it be expected that dicts can only have the direct indexing filter. Whichever you all think users would want and understand.

Reply:
Dicts insertion ordering is only guaranteed in python >= 3.7. Generally, range slicing of dicts is the extremely confusing feature and I would rather drop it. Key slicing `islice(["a", "b"])` is still useful and we should keep it.

Reply:
Noted. Will drop it then and throw an error instead for dicts if a range is provided